### PR TITLE
Dependency paths with multiple periods should resolve.

### PR DESCRIPTION
**Pull request by jimthedev**
Adds the .js extension back in the case of dependency paths with multiple periods in them.
Adds associated test.

### CODE REVIEW

File: test/test.js [hunk 1]
```diff
@@ -7,6 +7,14 @@ describe('resolve-dependency-path', function() {
     assert(resolved.indexOf(__dirname) === 0);
   });
 
+  it('resolves with multiple periods in the dependency path', function() {
+    var depPath = './bar.baz.qux';
```
Comment (mrjoelkemp) [inline test/test.js]: Hmm, how does this reach the else? Doesn't this technically have the ".qux" extension?

File: index.js [hunk 2]
```diff
@@ -24,6 +24,11 @@ module.exports = function(dep, filename, directory) {
 
   if (!depExt) {
     filepath += fileExt;
+  } else {
+    // If dep has multiple periods in the name add .js explicitly
+    if (fileExt === '.js') {
```
Comment (mrjoelkemp) [inline index.js]: What about 'index.js' as a path? Wouldn't that resolve to a duplicated extension?

**Comment by mrjoelkemp** [comment]
Good catch on this edge case. Thanks for contributing! 

Added a few questions to make sure we don't break things.

File: test/test.js (hunk 1 above)
Comment (jimthedev) [inline test/test.js]: Good q. It gets to the else block in the case of ./bar.baz.qux because path.extname always returns an empty string the string has a period as the first character (which is true in this case). http://nodejs.org/api/path.html#path_path_extname_p

Comment (mrjoelkemp) [inline test/test.js]: This is a bit more complicated than it seems. This module is also used for amd path resolutions (where loader plugins allow mustache, css, and text files as declared dependencies). For those types of dependencies, the `.js` extension will be incorrectly applied. 

Perhaps it's best to resolve the path, then check if a file with that path exists, and if not, find the key/extension in `require.extensions` that results in a found file (defaulting to `.js` if none were found to exist).

Comment (mrjoelkemp) [inline test/test.js]: Some additional tests (some of which are failing):

``` js
  describe('multiple period filenames', function() {
    it('resolves with multiple periods in the dependency path', function() {
      var depPath = './bar.baz.qux';
      var filename = __dirname + '/foo.js';
      var directory = __dirname;
      var resolved = resolvePath(depPath, filename, directory);
      assert(resolved === __dirname + '/bar.baz.qux.js');
    });

    it('does not duplicate extensions', function() {
      var depPath = '../index.js';
      var filename = __dirname + '/foo.js';
      var directory = __dirname;
      var resolved = resolvePath(depPath, filename, directory);
      // Extension after removing the .js extension
      var remainingExt = path.extname(path.basename(resolved, '.js'));
      assert.equal(remainingExt, '');
    });

    it('does not add the incorrect extension for sass files', function() {
      var depPath = 'styles';
      var filename = __dirname + '/foo.scss';
      var directory = __dirname;
      var resolved = resolvePath(depPath, filename, directory);
      assert.equal(path.extname(resolved), '.scss');
    });

    it('does not add the incorrect extension for mustache files', function() {
      var depPath = 'hgn!templates/foo.mustache';
      var filename = __dirname + '/foo.js';
      var directory = __dirname;
      var resolved = resolvePath(depPath, filename, directory);
      console.log(resolved)
      assert.equal(path.extname(resolved), '.mustache');
    });
  });
```

Comment (jimthedev) [inline test/test.js]: Interesting. Thanks for the additional tests. I will look at them. Unfortunately jspm actually does the opposite syntax as requirejs in terms of alternate loaders. It uses 'filename.txt!text' to denote the use of the text loader/plugin. :(

To the original question of the new else block. You are correct, my code broke two out of four of the following cases: './index.js', './index', 'index', and 'index.js'. To handle this, I've added tests for all four of these specific cases since they are very high impact cases. Shortly I will force push a new revision of my code with a modified else case. It should fix the two index.js cases that were previously breaking. Here are the tests I will add.

``` js
  it('resolves w/initial period, w/ending in .js', function() {
    var depPath = './index';
    var filename = __dirname + '/foo.js';
    var directory = __dirname;
    var resolved = resolvePath(depPath, filename, directory);
    assert(resolved === __dirname + '/index.js');
  });

  it('resolves w/initial period, w/o ending in .js', function() {
    var depPath = './index.js';
    var filename = __dirname + '/foo.js';
    var directory = __dirname;
    var resolved = resolvePath(depPath, filename, directory);
    assert(resolved === __dirname + '/index.js');
  });

  it('resolves w/o initial period, w/o ending in .js', function() {
    var depPath = 'index';
    var filename = __dirname + '/foo.js';
    var directory = __dirname;
    var resolved = resolvePath(depPath, filename, directory);
    assert(resolved === __dirname + '/index.js');
  });

  it('resolves w/o initial period, w/ending in .js', function() {
    var depPath = 'index.js';
    var filename = __dirname + '/foo.js';
    var directory = __dirname;
    var resolved = resolvePath(depPath, filename, directory);
    assert(resolved === __dirname + '/index.js');
  });
```

**Comment by jimthedev** [comment]
@mrjoelkemp Ok I included my new tests for index.js variants, systemjs dependency strings and those 4 tests that you mentioned in your comment. I don't know if there are downsides to assuming that having a ! in the dependency string means we don't need to throw .js on the end of that path. It seems to work in the current use-cases, but I don't know if there are other times when ! is used in a filename for other reasons. The conditional tree is getting a bit deep for my liking so if you have suggestions for refactoring that I'd appreciate it.

File: index.js [hunk 3]
```diff
@@ -1,4 +1,5 @@
 var path = require('path');
+require('string.prototype.endswith');
```
Comment (mrjoelkemp) [inline index.js:2]: Hmm, if `endsWith` is only used in one place, I don't see much point in pulling in the dependency. `depExt` already contains the extension, so we should just be able to use that.

File: test/test.js [hunk 4]
```diff
@@ -7,9 +8,40 @@ describe('resolve-dependency-path', function() {
     assert(resolved.indexOf(__dirname) === 0);
   });
 
+  it('resolves w/initial period, w/ending in .js', function() {
```
Comment (mrjoelkemp) [inline test/test.js:11]: Very thorough tests. This is great.

File: index.js [hunk 5]
```diff
@@ -24,6 +25,17 @@ module.exports = function(dep, filename, directory) {
 
   if (!depExt) {
     filepath += fileExt;
+  } else {
+    // If a dependency starts with a period AND it doesn't already end
+    // in .js AND doesn't use a custom plugin, add .js back to path.
+    if (fileExt === '.js' && !dep.endsWith('.js') && dep.indexOf('!') < 0) {
```
Comment (mrjoelkemp) [inline index.js:31]: It's great that you supplied systemjs support. This logic won't hold for webpack though, where `require('./styles/foo.css')` is a valid call within a JS file.

File: index.js [hunk 6]
```diff
@@ -24,6 +25,17 @@ module.exports = function(dep, filename, directory) {
...
+      filepath += fileExt;
+    } else {
+      // If using a SystemJS style plugin
+      if (depExt.indexOf('!') > -1) {
+        filepath += depExt.substring(0, depExt.indexOf('!'));
```
Comment (mrjoelkemp) [inline index.js:36]: This is really the resolution of a systemjs plugin path. You're saying that the systemjs loader will see `foo.css!` and resolve it to `foo.css`. This hints at a larger point where maybe plugin resolution should happen in this module. 

I had a separate one for the amd case: https://github.com/mrjoelkemp/node-module-lookup-amd and used that and this in the order here https://github.com/mrjoelkemp/node-module-lookup-amd:
1. Lookup the path (i.e., resolve aliases or plugin loading)
2. Resolve the dependency's path (to an absolute path)

The idea of "resolving a dependency's path" incorporates both steps really. What would be awesome is to have a module that takes in a dependency's path and returns the absolute path of that module on the filesystem.

Custom resolvers like sass could be hooked in based on the extension of the given filepath.

Thoughts?

I'm inclined to merge this and then follow up with the above functionality. It would save me a lot of redundant logic across a few modules.

**Comment by mrjoelkemp** [comment]
Landed in c41cd0069792a89c4fdb362d2a712e140cf9efc4 and 38e5321dcd323e1e3ad7a5b14eda79e16bae20bd
